account_mtd/models/mtd_hello_world.py

```diff
@@ -0,0 +1,406 @@
+        # Update the information in the api tracker table
+        self._logger.info("(Step 1) Get authorisation")
+        tracker_api = self.env['mtd.api_request_tracker']
+        tracker_api.create({
+            'user_id': self._uid,
+            'api_id': self.api_name.id,
+            'api_name': self.api_name.name,
+            'endpoint_id': self.id,
+            'request_sent': True,
+            })
+
+        redirect_uri = "{}/auth-redirect".format(self.hmrc_configuration.redirect_url)
+        authorisation_url = "https://test-api.service.hmrc.gov.uk/oauth/authorize?"
+        self._logger.info("(Step 1) Get authorisation - authorisation URI used:- {}".format(authorisation_url))
+        state = ""
+        # State is optional
+        if self.hmrc_configuration.state:
+            state = "&state={}".format(self.hmrc_configuration.state)
+        #scope needs to be percent encoded
+        import pdb; pdb.set_trace()
+        scope = urllib.parse.quote_plus(self.scope)
+
+        authorisation_url += (
+            "response_type=code&client_id={}&scope={}".format(self.hmrc_configuration.client_id, scope) +
+            "{}&redirect_uri={}".format(state, redirect_uri)
+        )
+        self._logger.info(
+            "(Step 1) Get authorisation - authorisation URI " +
+            "used to send request:- {}".format(authorisation_url)
+        )
+        req = requests.get(authorisation_url, timeout=3)
+        #response_token = json.loads(req.text)
+        self._logger.info(
+            "(Step 1) Get authorisation - received response of the request:- {}".format(req)
+        )
+        if req.ok:
+            return {'url': authorisation_url, 'type': 'ir.actions.act_url', 'target': 'self', 'res_id': self.id}
+        else:
+            response_token = json.loads(req.text)
+            construct_text = (
+                    "Date {}     Time {} \n".format(datetime.now().date(), datetime.now().time()) +
+                    "Sorry. The connection failed ! \n" +
+                    "Please check the log below for details. \n\n" +
+                    "Connection Status Details: \n" +
+                    "Request Sent: \n{} \n\n".format(authorisation_url) +
+                    "Error Code:\n{} \n\n".format(req.status_code) +
+                    "Response Received: \n{}".format(response_token['message'])
+            )
+            self.response_from_hmrc = construct_text
+            # We need to set the response received in tracker to be True 
+            # even if we receive negetive response.
+            tracker_api.response_received = True
+            action = self.env.ref('account_mtd.action_mtd_hello_world')
+            menu_id = self.env.ref('account_mtd.submenu_mtd_hello_world')
+            return werkzeug.utils.redirect(
+                '/web#id={}&view_type=form&model=mtd.hello_world'.format(self.id) +
+                '&menu_id={}&action={}'.format(menu_id.id, action.id))
```

Comment:
You may wish to look at https://docs.python.org/3/library/urllib.parse.html for functions specifically for building query strings and formatting URLs.  Not a huge biggy as long as all you're putting in are integers, but using string formatting will soon get very hairy as soon as other strings get involved.